api/src/main/java/io/druid/data/input/FirehoseFactory.java

```diff
@@ -36,7 +37,28 @@
    * If this method returns null, then any attempt to call hasMore(), nextRow(), commit() and close() on the return
    * value will throw a surprising NPE.   Throwing IOException on connection failure or runtime exception on
    * invalid configuration is preferred over returning null.
+   *
+   * @param parser             an input row parser
    */
-  public Firehose connect(T parser) throws IOException, ParseException;
+  @Deprecated
+  default Firehose connect(T parser) throws IOException, ParseException
+  {
+    return connect(parser, null);
+  }
 
+  /**
+   * Initialization method that connects up the fire hose.  If this method returns successfully it should be safe to
+   * call hasMore() on the returned Firehose (which might subsequently block).
+   * <p/>
+   * If this method returns null, then any attempt to call hasMore(), nextRow(), commit() and close() on the return
+   * value will throw a surprising NPE.   Throwing IOException on connection failure or runtime exception on
+   * invalid configuration is preferred over returning null.
+   * <p/>
+   * Some fire hoses like {@link PrefetchableTextFilesFirehoseFactory} may use a temporary
+   * directory to cache data in it.
+   *
+   * @param parser             an input row parser
+   * @param temporaryDirectory a directory where temporary files are stored
+   */
+  Firehose connect(T parser, File temporaryDirectory) throws IOException, ParseException;
```

Comment:
This one should have a default implementation too (for compatibility with pre-existing firehose impls that don't know about overriding this method). It's ok for them to both have default impls that call each other -- anyone implementing a firehose should override at least one. (it's worth calling this out in the javadoc)